src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs

```diff
@@ -86,11 +86,17 @@ public CommonCompiler(CommandLineParser parser, string responseFile, string[] ar
 
         internal abstract bool SuppressDefaultResponseFile(IEnumerable<string> args);
 
-        internal string GetAssemblyFileVersion()
+        /// <summary>
+        /// The type of the compiler class for version information in /help and /version.
+        /// We don't simply use this.GetType() because that would break mock subclasses.
+        /// </summary>
+        internal abstract Type Type { get; }
+
+        internal virtual string GetAssemblyFileVersion()
         {
             if (_clientDirectory != null)
             {
-                var name = $"{typeof(CommonCompiler).GetTypeInfo().Assembly.GetName().Name}.dll";
+                var name = $"{Type.GetTypeInfo().Assembly.GetName().Name}.dll";
```

Comment:
Strictly the name of the assembly and the file could differ. Is there no filename or location option you can use?